### ISSUE TITLE
Occasional Segmentation fault on threaded MSEED-reading from SDS-client

### ISSUE THREAD

**Issue by flixha** [obspy/obspy#3114]
### Avoid duplicates

- [X] I searched existing issues

### Bug Summary

I'm trying to use the function `client.get_waveforms` for an SDS-archive in parallel, using a `ThreadPool` that is a lot quicker for this Disk-bound task than a Process-Pool (see code below). Generally this works really well, but there are some files in the SDS archive where obspy occasionally (i.e., not always) quits with a segmentation fault. As a SegFault can't be caught in Python, this is very annoying when running in larger programs.

- I have tried to narrow the occurrence of the segmentation fault down to a minimal example (quite a tricky task because the fault doesn't always occur and depends on at least 4-5 files being present in SDS archive).

- To run the minimal sample below, here's the smallest SDS archive that I could reproduce it with (untar with `tar xvf SLARCHIVE.tar.gz`, put into folder where example is run): [SLARCHIVE.tar.gz](https://github.com/obspy/obspy/files/9207247/SLARCHIVE.tar.gz).

- Reproducing the segfault may require a few tries (with this minimal example it appears to be rarer than with larger requests), you can try with: `for j in {1..10}; do python debug_segfault_min_pub.py; echo "\n\n"; done`. I have reproduced it on two different systems.

I was not able to reproduce the segfault with:
- reading in serial
- using `obspy.read()` alone in parallel (i.e., needs SDS-client)
- using a pool of processes (sounds like a nice workaround, but is 5x slower...)

Any ideas on how / whether we can fix this? I'm a bit lost on how to continue debugging / fixing this... (I found two possibly related issues / PRs from a while ago: https://github.com/obspy/obspy/pull/1658, https://github.com/obspy/obspy/pull/1728) @krischer, @QuLogic , @megies)

### Code to Reproduce

```python
import logging
Logger = logging.getLogger(__name__)
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s\t%(name)40s:%(lineno)s\t%(funcName)20s()\t%(levelname)s\t%(message)s")
"[%(filename)s:%(lineno)s - %(funcName)20s() ] %(message)s"

from obspy import Stream, read
from multiprocessing.pool import ThreadPool


def get_parallel_waveform_client(waveform_client):
    """
    Bind a `get_waveforms_bulk` method to waveform_client if it doesn't already
    have one.
    """
    def _get_waveforms_bulk_parallel_naive(self, bulk, parallel=True):
        """
        parallel implementation of get_waveforms_bulk.
        """
        st = Stream()
        if parallel:
            pool = ThreadPool(processes=2)
            results = [pool.apply_async(self.get_waveforms, args=arg)
                       for arg in bulk]
            pool.close()
            pool.join()
            # Need to handle possible read-errors in each request when getting
            # each request-result.
            for res in results:
                try:
                    st += res.get()
                except Exception as e:
                    Logger.error(e)
                    pass
        else:
            for arg in bulk:
                try:
                    st += self.get_waveforms(*arg)
                except Exception as e:
                    Logger.error(e)
                    continue
        return st

    # add waveform_bulk method dynamically if it doesn't exist already
    if not hasattr(waveform_client, "get_waveforms_bulk_parallel"):
        bound_method = _get_waveforms_bulk_parallel_naive.__get__(
            waveform_client)
        setattr(waveform_client, "get_waveforms_bulk_parallel", bound_method)

    return waveform_client


# %%

from obspy.core import UTCDateTime
from obspy.clients.filesystem.sds import Client

client = Client('SLARCHIVE')
client = get_parallel_waveform_client(client)

bulk_request = [
    ('*', 'KMY', '*', 'HHE', UTCDateTime(2019,1,31), UTCDateTime(2019,1,31,23,59,59)),
    ('*', 'KMY', '*', 'HHN', UTCDateTime(2019,1,31), UTCDateTime(2019,1,31,23,59,59))]

Logger.info('Requesting waveforms from client %s', client)
for j in range(0,30):
    day_st = client.get_waveforms_bulk_parallel(bulk_request, parallel=True)
    Logger.info('Parallel read: success %s', j)

Logger.info('All done successfully in parallel')
```

### Error Traceback

```Python traceback
Segmentation faults, so no really useful traceback. I have seen 3 different types of exits:

- Trace/breakpoint trap (core dumped)
- Segmentation fault (core dumped)
- Illegal instruction (core dumped)

The program appears to segfault on these lines when reading SLARCHIVE/2019/NS/KMY/HHN.D/NS.KMY.00.HHN.D.2019.031: https://github.com/obspy/obspy/blob/4fcb7687d78afc6157d2dd15c057cdeee0a5ffdf/obspy/io/mseed/core.py#L395-L398

This same file also gives the following warning:
InternalMSEEDWarning: NS_KMY_00_HHN_D: Warning: Data integrity check for Steim2 failed, Last sample=-39588, Xn=-39576
```

### ObsPy Version?

1.3.0.post0+271.g4fcb7687d7 (master)

### Operating System?

RHEL 8.6, CentOS 7

### Python Version?

3.9.12

### Installation Method?

developer installation / from source / git checkout

**Comment by calum-chamberlain** [comment]
Just a +1 from me (sorry for no help on this yet): I have run into seg-faults when reading mseed files using a threadpool (without an SDS client) as well that I can't replicate in serial. It similarly seems to happen when there are files that fail the data integrity check for Steim2 that @flixha mentioned.

**Comment by flixha** [comment]
I suspect that the SDS client shouldn't cause the problems as it doesn't really touch functions in C that probably cause the segfault. But to cause the segfault, it seems that there needs to be a specific combination and / or order of mseed-files and starttime / endtime that are sent to the C-functions that sometimes cause the segfault; and I wasn't able to isolate that yet. Maybe it depends on the order in which the threads access some memory?

It would of course be good to narrow this down more; if you have 2 files that alone can cause the segfault that would be nice to have. I'm also trying to produce a useful coredump; but I'm still lacking some debug symbols in my attempts to have the C-functions and python-dbg compiled with debug symbols.

**Comment by megies** [comment]
I can confirm the problem with your code and minimal SDS file tree.

Removing both files `*.2019.032` (which do not contain any requested data) still shows the issue for me, reducing the file pool to 3 files, with the one `*2019.030` file containing a short sliver of wanted data (3 sec).

```
SLARCHIVE/
└── 2019
    └── NS
        └── KMY
            ├── HHE.D
            │   ├── NS.KMY.00.HHE.D.2019.030
            │   └── NS.KMY.00.HHE.D.2019.031
            └── HHN.D
                └── NS.KMY.00.HHN.D.2019.031
```

Reducing it even further, I can still get `Illegal instruction` and `segmentation fault` **very** rarely (like 1 out of 40 tries) when reducing the file `*.030` to the last record (i.e. the last 512 bytes), which is the only record containing requested data.

Maybe @chad-iris has an idea what could be a problem here, since it likely will have to do with libmseed?
